### PR TITLE
feat: AddRingBuf() using ring_buffer__add

### PR DESCRIPTION
**Pull request by mighty1231**
libbpf provides ring_buffer__add to handle multiple ringbuf event callbacks

- single RingBuffer object now handles multiple slots
- implements AddRingBuf() for ring_buffer__add wrapper
- updates ringbuffers selftest to show how to handle it

Reference
- https://github.com/torvalds/linux/blob/eb6a9339efeb6f3d2b5c86fdf2382cdc293eca2c/tools/testing/selftests/bpf/progs/test_ringbuf_multi.c
- https://github.com/torvalds/linux/blob/eb6a9339efeb6f3d2b5c86fdf2382cdc293eca2c/tools/testing/selftests/bpf/prog_tests/ringbuf_multi.c#41

### CODE REVIEW

**Comment by geyslan** [comment]
@mighty1231 tks for this PR. There's a formatting issue https://github.com/aquasecurity/libbpfgo/pull/430/checks. As it is solved I'll stash it for later review. Tks again.

**Comment by mighty1231** [comment]
@geyslan I ran `make fmt-fix`. Now `make fmt-check` returns no error

**Comment by mighty1231** [comment]
Failed test `map-getmapsbyname` seems not related to my PR, because it seems don't use any ringbuffer, and it runs before the `ringbuffers` test.

**Comment by geyslan** [comment]
@mighty1231, I suspect that self-test/iter sometimes fails due to ping usage and internal network malfunction. Well, it seems like a flaky test.

**Review by geyslan — CHANGES_REQUESTED** [review]
@mighty1231 I tested it and it works nice. That's awesome.

I would just ask you to make the selftest more robust, since you have touched the logic of the slot (index) inside RingBuffer that's used by `rwArray` which .

Would you mind to improve the the respective selftest to submit a greater volume of data through a greater number of slots (ringbuffers) not reusing the same data holder? I mean, try using more than one program to submit the data.

**Comment by mighty1231** [comment]
> @mighty1231 I tested it and it works nice. That's awesome.
> 
> I would just ask you to make the selftest more robust, since you have touched the logic of the slot (index) inside RingBuffer that's used by `rwArray` which .
> 
> Would you mind to improve the the respective selftest to submit a greater volume of data through a greater number of slots (ringbuffers) not reusing the same data holder? I mean, try using more than one program to submit the data.

I am confused. What did you mean among following tests?

1. More tests to check thread safety of `Module *AddRingBuf()`, `RingBuffer *Stop()`, and `RingBuffer *Close()`
2. More tests with two or more `struct *RingBuffer` in golang
3. More tests wtih two or more `NewModuleFromFile("main.bpf.o")` call
4. More tests wtih two or more probes - other than `kprobe__sys_mmap` from `selftest/ringbuffers/main.bpf.c`

1 may be resolved with notifying the limitation for users. 2 - The motivation of this PR. I have got completely lost to use more than one RingBuffer, so I decided to use `ring_buffer__add`. And, it's difficult to guess what can be improved by case 3 and 4

File: libbpfgo.c [hunk 1]
```diff
@@ -59,6 +59,17 @@ struct ring_buffer *cgo_init_ring_buf(int map_fd, uintptr_t ctx)
     return rb;
 }
 
+int cgo_add_ring_buf(struct ring_buffer *rb, int map_fd, uintptr_t ctx)
+{
+    int ret = ring_buffer__add(rb, map_fd, ringbufferCallback, (void *) ctx);
+    if (ret != 0) {
+        fprintf(stderr, "Failed to add ring buffer: %s\n", strerror(-ret));
+        return ret;
```
[the remark this answers is not in the document]
Comment (mighty1231) [reply libbpfgo.c]: I changed it.

**Comment by geyslan** [comment]
>     5. Make similar to ringbuf_multi test from linux repository

It would be awesome to have a huge amount of tests like ringbuf_multi. Anyway, I think it would take longer.

I meant:

- increasing `numberOfEvent*Received*` to a considerable amount.
- split the `recvLoop` `for` `select` into the amount of ringbuf slots, raising one goroutine for each.

**Comment by mighty1231** [comment]
I don't know how to split the for-select loop, and I didn't understand what is improved with new test.

**Comment by geyslan** [comment]
@NDStrahilevitz do you mind reviewing this?

**Comment by geyslan** [comment]
@mighty1231 I'm merging this. Thanks again for your contrib.

**Pull request merged**